File: src/core/markdown.ts
export function cleanMarkdownResponse(response: string): string {
  console.debug('cleanMarkdownResponse input length:', response.length);

  let cleaned = response.trim();

  cleaned = cleaned.replace(/<think\b[^>]*>[\s\S]*?<\/think>/gi, '');
  cleaned = cleaned.replace(/<thinking\b[^>]*>[\s\S]*?<\/thinking>/gi, '');

  if (cleaned.indexOf('\n---\n') === -1) {
    const headerMatch = cleaned.match(/\n#{1,2} \S/);
    if (headerMatch) {
      const cutIdx = cleaned.indexOf(headerMatch[0]);
      if (cutIdx > 0) {
        cleaned = cleaned.slice(cutIdx + 1).replace(/^\s+/, '');
      }
    }
  }

  const codeBlockPatterns = [
    /^```(?:markdown|md)?\s*\n([\s\S]*?)\n```$/gm,
    /^```(?:markdown|md)?\s*([\s\S]*?)```$/gm,
    /^```(?:markdown|md)?\s*\n([\s\S]*)$/gm,
    /^```(?:markdown|md)?\s*([\s\S]*)$/gm,
  ];

  for (const pattern of codeBlockPatterns) {
    const match = cleaned.match(pattern);
    if (match) {
      cleaned = cleaned.replace(pattern, '$1').trim();
      console.debug('code block wrapping detected, removed');
      break;
    }
  }

  if (cleaned.startsWith('```')) {
    cleaned = cleaned.replace(/^```(?:markdown|md)?\s*\n?/, '');
    console.debug('removed opening code block marker');
  }

  if (cleaned.endsWith('```')) {
    cleaned = cleaned.replace(/\n?```$/, '');
    console.debug('removed closing code block marker');
  }

  console.debug('cleanMarkdownResponse output length:', cleaned.length);
  console.debug('first 50 chars:', cleaned.substring(0, 50));

  if (!cleaned.startsWith('---')) {
    const fmEnd = cleaned.indexOf('\n---\n');
    if (fmEnd !== -1) {
      const beforeFm = cleaned.substring(0, fmEnd);
      const looksLikeFrontmatter =
        beforeFm.includes(':') &&
        !beforeFm.startsWith('#') &&
        !beforeFm.startsWith('```') &&
        beforeFm.split('\n').filter(l => l.trim()).every(l => l.includes(':') || l.trim() === '');
      if (looksLikeFrontmatter) {
        cleaned = '---\n' + cleaned;
        console.debug('added missing opening ---');
      } else {
        cleaned = cleaned.substring(fmEnd + 1);
        console.debug('removed preamble text before frontmatter');
      }
    }
  }

  return cleaned.trim();
}

/**
 * Extract reasoning blocks from LLM response content.
 *
 * v1.20.0: When the user runs thinking-capable models with the default
 * "no explicit thinking control" mode, providers may emit reasoning as:
 *   - <think>...</think> blocks
 *   - <thinking>...</thinking> blocks (alternative XML)
 *   - <think role="assistant">...</think> (with attributes)
 *
 * The Query Wiki UI renders the visible content normally and shows the
 * reasoning inside a <details> collapsible block (default collapsed), so
 * reasoning never visually leaks into the main answer. This extractor
 * normalizes all three forms into a single shape.
 *
 * Pure function — no side effects.
 *
 * @param content - Raw LLM response text
 * @returns Object with `thinkingBlocks: string[]` (in order) and
 *          `visibleContent: string` (with the blocks removed)
 */
export function extractThinkingBlocks(content: string): {
  thinkingBlocks: string[];
  visibleContent: string;
} {
  if (!content) {
    return { thinkingBlocks: [], visibleContent: '' };
  }

  // Match both <think>...</think> and <thinking>...</thinking> with optional
  // attributes on the opening tag. Use non-greedy [\s\S]*? so multiple blocks
  // are matched individually rather than as one giant block.
  // Require a closing tag — incomplete blocks (no close) are left in the
  // visible content to avoid swallowing the rest of the response.
  const blockRegex = /<think(?:ing)?\b[^>]*>[\s\S]*?<\/think(?:ing)?>\s*/gi;
  const thinkingBlocks: string[] = [];
  let visibleContent = content.replace(blockRegex, (_match) => {
    return '';
  });

  // Re-extract the inner content of each block for the result. Doing the
  // replacement in two passes (strip with whitespace, then collect inner)
  // is simpler than passing through a callback with capture groups.
  const innerRegex = /<think(?:ing)?\b[^>]*>([\s\S]*?)<\/think(?:ing)?>/gi;
  let m: RegExpExecArray | null;
  while ((m = innerRegex.exec(content)) !== null) {
    thinkingBlocks.push(m[1].trim());
  }

  return { thinkingBlocks, visibleContent: visibleContent.trimStart() };
}

/**
 * Encode reasoning_content into a <think> block and prepend to visible text.
 * This is the encoding counterpart to extractThinkingBlocks() — the two
 * functions share the same `<think>...</think>` delimiter contract.
 *
 * @param reasoning - Raw reasoning content (e.g. DeepSeek reasoning_content)
 * @param text - Visible response text
 * @returns Combined string with reasoning wrapped in <think> tags, or just text
 */
export function wrapReasoningContent(reasoning: string, text: string): string {
  if (!reasoning) return text;
  return '<think>' + reasoning + '</think>\n\n' + text;
}

File: python_lib/llm_wiki/core/markdown.py
"""
Markdown utilities — port of src/core/markdown.ts
"""
from __future__ import annotations

import re


def clean_markdown_response(response: str) -> str:
    """Strip LLM artifacts from a generated markdown page."""
    cleaned = response.strip()

    # Remove thinking blocks
    cleaned = re.sub(r"<think\b[^>]*>[\s\S]*?</think>", "", cleaned, flags=re.IGNORECASE)
    cleaned = re.sub(r"<thinking\b[^>]*>[\s\S]*?</thinking>", "", cleaned, flags=re.IGNORECASE)

    # Try to find the real frontmatter / header start
    if "\n---\n" not in cleaned:
        header_match = re.search(r"\n#{1,2} \S", cleaned)
        if header_match:
            cut_idx = cleaned.index(header_match.group())
            if cut_idx > 0:
                cleaned = cleaned[cut_idx + 1:].lstrip()

    # Strip code-block wrappers
    code_block_patterns = [
        r"^```(?:markdown|md)?\s*\n([\s\S]*?)\n```$",
        r"^```(?:markdown|md)?\s*([\s\S]*?)```$",
        r"^```(?:markdown|md)?\s*\n([\s\S]*)$",
        r"^```(?:markdown|md)?\s*([\s\S]*)$",
    ]
    for pattern in code_block_patterns:
        m = re.match(pattern, cleaned, re.MULTILINE)
        if m:
            cleaned = m.group(1).strip()
            break

    if cleaned.startswith("```"):
        cleaned = re.sub(r"^```(?:markdown|md)?\s*\n?", "", cleaned)
    if cleaned.endswith("```"):
        cleaned = re.sub(r"\n?```$", "", cleaned)

    # Fix missing opening ---
    if not cleaned.startswith("---"):
        fm_end = cleaned.find("\n---\n")
        if fm_end != -1:
            before_fm = cleaned[:fm_end]
            looks_like_fm = (
                ":" in before_fm
                and not before_fm.startswith("#")
                and not before_fm.startswith("```")
                and all(l.strip() == "" or ":" in l for l in before_fm.split("\n") if l.strip())
            )
            if looks_like_fm:
                cleaned = "---\n" + cleaned
            else:
                cleaned = cleaned[fm_end + 1:]

    return cleaned.strip()


def extract_thinking_blocks(content: str) -> dict:
    """
    Extract reasoning blocks from LLM response content.
    Returns dict with 'thinking_blocks' (list) and 'visible_content' (str).
    """
    if not content:
        return {"thinking_blocks": [], "visible_content": ""}

    block_regex = re.compile(
        r"<think(?:ing)?\b[^>]*>[\s\S]*?</think(?:ing)?>\s*",
        re.IGNORECASE,
    )
    visible_content = block_regex.sub("", content)

    inner_regex = re.compile(
        r"<think(?:ing)?\b[^>]*>([\s\S]*?)</think(?:ing)?>",
        re.IGNORECASE,
    )
    thinking_blocks = [
        unescape_thinking_tag(m.group(1).strip())
        for m in inner_regex.finditer(content)
    ]

    return {
        "thinking_blocks": thinking_blocks,
        "visible_content": visible_content.lstrip(),
    }


def wrap_reasoning_content(reasoning: str, text: str) -> str:
    """Encode reasoning_content into a <think> block prepended to text."""
    if not reasoning:
        return text
    safe_reasoning = re.sub(r"</think", r"<\\/think", reasoning, flags=re.IGNORECASE)
    return f"<think>{safe_reasoning}</think>\n\n{text}"


def unescape_thinking_tag(s: str) -> str:
    return re.sub(r"<\\/think", "</think", s, flags=re.IGNORECASE)
